### PR TITLE
Adds enable/disable PDF export via config

### LINKED ISSUE #561: Disable PDF export by config
The PDF export is currently experimental and causes problems when running on Alpine Linux or other musl-based distros because phantomjs doesn't work.

By allow to disable this feature by a config value, the alpine build of https://github.com/hackmdio/docker-hackmd/pull/14 will no longer break.

### PR DESCRIPTION
**Pull request by geekyd**
Adds a config variable for enabling & disabling PDF exports

Fixes #561

### CODE REVIEW

**Comment by SISheogorath** [comment]
👍  for this PR!

Please have a look at:

https://github.com/hackmdio/hackmd/blob/5ce8f40eac7fccfcfca175cf79ccfa86a84bc113/lib/response.js#L285-L307

and

https://github.com/hackmdio/hackmd/blob/5ce8f40eac7fccfcfca175cf79ccfa86a84bc113/lib/response.js#L384-L386

We need a server-side implementation because this export function uses PhantomJS which doesn't work on [musl libc](https://www.musl-libc.org/), which is the library for alpine. To make a long story short, when you run the pdf export on the alpine based container, HackMD crashes (DoS attack vector) and we want to prevent this.

Can you please have a look at this with this PR, too?

**Comment by geekyd** [comment]
@SISheogorath Thanks for the quick response. I will take a look at it. Just to be clear about the requirements here, we can go for something like this:

```
 function actionPDF (req, res, note) { 
   var body = note.content 
   var extracted = models.Note.extractMeta(body) 
   var title = models.Note.decodeTitle(note.title) 
  
   if (!fs.existsSync(config.tmppath)) { 
     fs.mkdirSync(config.tmppath) 
   } 
   var path = config.tmppath + '/' + Date.now() + '.pdf' 

   if(config.allowPDFExport) {
   markdownpdf().from.string(extracted.markdown).to(path, function () { 
     var stream = fs.createReadStream(path) 
     var filename = title 
     // Be careful of special characters 
     filename = encodeURIComponent(filename) 
     // Ideally this should strip them 
     res.setHeader('Content-disposition', 'attachment; filename="' + filename + '.pdf"') 
     res.setHeader('Cache-Control', 'private') 
     res.setHeader('Content-Type', 'application/pdf; charset=UTF-8') 
     res.setHeader('X-Robots-Tag', 'noindex, nofollow') // prevent crawling 
     stream.pipe(res) 
     fs.unlink(path) 
   }) 
  }
 } 
```

OR 

```
case 'pdf': 
   if(config.allowPDFExport) {
    actionPDF(req, res, note) 
  }
   break 
```
Please correct me if I'm wrong

**Comment by SISheogorath** [comment]
After thinking about it, the last method looks better to me. but we should handle it correctly and return 403 as error code when the function is disabled.

**Comment by SISheogorath** [comment]
Also please sign https://github.com/hackmdio/hackmd/pull/578 so we can make sure this and future changes are compatible with the new AGPL-3.0 license.

It would help us a lot to keep HackMD's CE awesome ;)

**Comment by geekyd** [comment]
@SISheogorath I have signed the license change ([here](https://github.com/hackmdio/hackmd/pull/578#issuecomment-336739337))

File: lib/response.js [hunk 1]
```diff
@@ -382,7 +384,11 @@ function noteActions (req, res, next) {
         actionInfo(req, res, note)
         break
       case 'pdf':
-        actionPDF(req, res, note)
+        if (config.allowPDFExport) {
+          actionPDF(req, res, note)
+        } else {
+          response.errorForbidden(res)
```
Comment (SISheogorath) [inline lib/response.js]: We should also add `logger.error('PDF export failed: Disabled by config. Set "allowPDFExport: true" to enable. Check the documentation for details')` so no one is confused and we provide a helpful error message.

Comment (geekyd) [reply lib/response.js]: @SISheogorath  Thanks for the suggestion. Added the error message

**Comment by geekyd** [comment]
@SISheogorath Updated! Sorry for the delay.

**Comment by SISheogorath** [comment]
@geekyd no problem. But please check your latest changes. -> https://github.com/hackmdio/hackmd/pull/589/files#diff-2ec1599c37878db948302f7e8152abf4L35

With this PR you would undo https://github.com/hackmdio/hackmd/pull/595

**Comment by geekyd** [comment]
@SISheogorath Updated. :)

**Comment by SISheogorath** [comment]
Thanks a lot!

**Pull request merged**